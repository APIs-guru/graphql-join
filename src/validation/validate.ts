import {
  GraphQLError,
  DocumentNode,

  validate,

  ArgumentsOfCorrectTypeRule,
  KnownArgumentNamesRule,
  KnownDirectivesRule,
  ProvidedNonNullArgumentsRule,
  UniqueArgumentNamesRule,
  UniqueDirectivesPerLocationRule,
  UniqueInputFieldNamesRule,
} from 'graphql';

import { directiveSchema } from '../directives';

export function validateJoinIDL(ast: DocumentNode): Array<GraphQLError> {
  return validateDirectives(ast);
}

function validateDirectives(ast: DocumentNode): Array<GraphQLError> {
  // FIXME: should allow directives from sendTo remote schema on operations
  // FIXME: check that there no query arguments inside directive values
  return validate(directiveSchema, ast, [
    ArgumentsOfCorrectTypeRule,
    KnownArgumentNamesRule,
    KnownDirectivesRule,
    ProvidedNonNullArgumentsRule,
    UniqueArgumentNamesRule,
    UniqueDirectivesPerLocationRule,
    UniqueInputFieldNamesRule,
  ]);
  // TODO:
  // JOIN AST:
  //   - validate that all directive known and locations are correct
  //   - type refs should be resolved without conflicts
  //   - no specified directives inside join AST
  //   - all references to remote types have no conficts
  //   - all fields inside extends and type defs should have @resolveWith
  //   - all field args + all fragment exports used in operation
  //   - fields return types should match types returned by "query" including LIST and NON_NULL
  //   - TEMPORARY: @resolveWith for root fields shouldn't use fragments
  // schema:
  //   - check for subscription in schema and `Subscription` type and error as not supported
  //   - references to Query and Mutation roots point to exactly one remote type
  //   - reference only roots of the same type from remote schema
  // fragments:
  //   - shoud have uniq names
  //   - shouldn't reference other fragments
  //   - should have atleast one leaf
  //   - all scalars should have exports directive
  //   - names in export directive should be uniq
  //   - should be used in @resolveWith
  //   - no field alliases
  //   - forbid @skip/@include
  //   - TEMPORARY: fragment shouldn't contain fields with @resolveWith
  //   - should be used only on objects which is equal or inherate from type condition
  // operations:
  //   - only query and mutation no subscription
  //   - mutation should be used only on fields in mutation root
  //   - should have name
  //   - shoud have uniq names
  //   - should have @send(to:)
  //   - valid against external schema
  //   - should have exactly one "leaf" which is "{...CLIENT_SELECTION}" or scalar
  //   - don't reference other fragments
  //   - should be used in @resolveWith
  //   - no field alliases
  //   - forbid @skip/@include
}
