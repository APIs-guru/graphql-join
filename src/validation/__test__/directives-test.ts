import { parse } from 'graphql';
import { validateJoinIDL } from '../validate';

function snapshotsErrors(idl: string) {
  expect(validateJoinIDL(parse(idl))).toMatchSnapshot();
}

describe('validate directives', () => {
  test('unknown directives', () => {
    snapshotsErrors(`
      type Query {
        foo: String @unknown
      }
    `);
  });

  test('missing required argument', () => {
    snapshotsErrors(`
      type Query {
        foo: String @resolveWith
      }
    `);
  });
});
