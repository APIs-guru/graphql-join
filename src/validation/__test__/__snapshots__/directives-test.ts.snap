// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate directives missing required argument 1`] = `
Array [
  [GraphQLError: Directive "@resolveWith" argument "query" of type "String!" is required but not provided.],
]
`;

exports[`validate directives unknown directives 1`] = `
Array [
  [GraphQLError: Unknown directive "unknown".],
]
`;
